File: examples/Calculator.hs
{-# LANGUAGE TypeOperators
           , FlexibleContexts #-}
module Calculator where

import Control.Monad

import SessionCheck
import SessionCheck.Backend.Erlang

calculator :: (String :< t, Int :< t) => Spec t ()
calculator = do
  op <- choose ["mul", "div", "stop"]
  case op of
    "mul" -> do
      i <- send anyInt
      j <- send anyInt
      get $ is (i * j)
    "div" -> do
      i <- send nonNegInt
      j <- send posInt
      get $ is (i `div` j)
    "stop" -> stop
  calculator

main :: IO ()
main = do
  checkCoherence calculator
  erlangMain "calculator:main" calculator 

File: examples/calculator.erl
-module(calculator).
-export([main/0]).

main() -> spawn(fun() -> f() end).

f() ->
  register(p, self()),
  receive 
    _ -> loop()
  end.

getInt() ->
  receive
    {_, V} -> V
  end.

loop() ->
  receive 
    {_ , "stop"} -> exit(done);
    {HS, "mul"} ->
          I = getInt(),
          J = getInt(),
          HS ! (I * J);
    {HS, "div"} -> 
          I = getInt(),
          J = getInt(),
          HS ! (I div J)
  end,
  loop().
